File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/model/Label.java
package org.jetbrains.bsp.bazel.server.sync.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Objects;

public class Label {
  private final String value;

  public Label(@JsonProperty("value") String value) {
    this.value = value;
  }

  public static Label from(String value) {
    return new Label(value);
  }

  public String getValue() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Label that = (Label) o;
    return Objects.equals(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return value;
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/model/Label.kt
package org.jetbrains.bsp.bazel.server.model

import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.KeyDeserializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import com.fasterxml.jackson.databind.ser.std.StdScalarSerializer
import com.fasterxml.jackson.databind.ser.std.StdSerializer

@JvmInline
value class Label private constructor(@JsonValue val value: String) {

  fun targetName(): String =
    value.substringAfterLast(":", "")

  override fun toString(): String = value

  companion object {
    fun parse(value: String): Label =
      Label(value.intern())
  }
}

class LabelKeyDeserializer : KeyDeserializer() {
  override fun deserializeKey(key: String, ctxt: DeserializationContext): Label {
    return Label.parse(key)
  }
}
